File: app/src/main/java/com/bijoysingh/quicknote/activities/ViewAdvancedNoteActivity.java
package com.bijoysingh.quicknote.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;

import com.bijoysingh.quicknote.R;
import com.bijoysingh.quicknote.activities.sheets.NoteAdvancedActivityBottomSheet;
import com.bijoysingh.quicknote.activities.sheets.NoteSettingsOptionsBottomSheet;
import com.bijoysingh.quicknote.activities.sheets.TextSizeBottomSheet;
import com.bijoysingh.quicknote.database.Note;
import com.bijoysingh.quicknote.database.Tag;
import com.bijoysingh.quicknote.formats.Format;
import com.bijoysingh.quicknote.formats.FormatType;
import com.bijoysingh.quicknote.recyclerview.FormatAdapter;
import com.bijoysingh.quicknote.recyclerview.FormatTextViewHolder;
import com.bijoysingh.quicknote.utils.CircleDrawable;
import com.bijoysingh.quicknote.utils.NoteState;
import com.bijoysingh.quicknote.utils.ThemeColorType;
import com.bijoysingh.quicknote.utils.ThemeManager;
import com.github.bijoysingh.starter.prefs.DataStore;
import com.github.bijoysingh.starter.recyclerview.RecyclerViewBuilder;
import com.google.android.flexbox.FlexboxLayout;

import java.util.List;
import java.util.Set;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.bijoysingh.quicknote.activities.sheets.SettingsOptionsBottomSheet.KEY_MARKDOWN_ENABLED;
import static com.bijoysingh.quicknote.utils.NoteBuilderKt.genEmptyNote;
import static com.bijoysingh.quicknote.utils.ThemeManagerKt.KEY_NIGHT_THEME;

public class ViewAdvancedNoteActivity extends ThemedActivity {

  public static final String NOTE_ID = "NOTE_ID";
  public static final int HANDLER_UPDATE_TIME = 1000;

  protected Context context;
  protected Note note;
  protected DataStore store;

  protected FormatAdapter adapter;
  protected List<Format> formats;

  public Format focusedFormat;

  protected View toolbar;
  protected RecyclerView formatsView;

  protected View rootView;
  protected ImageView backButton;
  protected ImageView actionCopy;
  protected ImageView actionDelete;
  protected ImageView actionShare;
  protected ImageView actionEdit;
  protected ImageView actionDone;
  protected ImageView actionOptions;
  protected FlexboxLayout colorSelectorLayout;
  protected ImageView colorButton;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_advanced_note);
    context = this;
    store = DataStore.get(context);

    int noteId = getIntent().getIntExtra(NOTE_ID, 0);
    if (noteId == 0 && savedInstanceState != null) {
      noteId = savedInstanceState.getInt(NOTE_ID, 0);
    }
    if (noteId != 0) {
      note = Note.db(this).getByID(noteId);
    }
    note = note == null
        ? genEmptyNote(NoteSettingsOptionsBottomSheet.Companion.genDefaultColor(store))
        : note;

    rootView = findViewById(R.id.root_layout);
    setRecyclerView();
    setToolbars();
    setEditMode();
    notifyThemeChange();
  }

  public static Intent getIntent(Context context, Note note) {
    Intent intent = new Intent(context, ViewAdvancedNoteActivity.class);
    intent.putExtra(NOTE_ID, note.uid);
    return intent;
  }

  @Override
  protected void onResume() {
    super.onResume();
    onResumeAction();
  }

  protected void onResumeAction() {
    note = Note.db(this).getByID(getIntent().getIntExtra(NOTE_ID, 0));
    if (note == null) {
      finish();
      return;
    }
    setNote();
  }

  protected void setEditMode() {
    setEditMode(getEditModeValue());
    formatsView.setBackgroundColor(ThemeManager.Companion.get(this).get(this, ThemeColorType.BACKGROUND));
  }

  protected boolean getEditModeValue() {
    return false;
  }

  protected void setEditMode(boolean mode) {
    resetBundle();
    setNote();

    actionEdit.setVisibility(mode ? GONE : VISIBLE);
    actionDone.setVisibility(mode ? VISIBLE : GONE);
    toolbar.setVisibility(mode ? VISIBLE : GONE);
  }

  private void resetBundle() {
    Bundle bundle = new Bundle();
    bundle.putBoolean(FormatTextViewHolder.Companion.getKEY_EDITABLE(), getEditModeValue());
    bundle.putBoolean(KEY_MARKDOWN_ENABLED, store.get(KEY_MARKDOWN_ENABLED, true));
    bundle.putBoolean(KEY_NIGHT_THEME, ThemeManager.Companion.get(this).isNightTheme());
    bundle.putInt(TextSizeBottomSheet.KEY_TEXT_SIZE, TextSizeBottomSheet.Companion.getDefaultTextSize(store));
    adapter.setExtra(bundle);
  }

  protected void setNote() {
    setNoteColor(note.color);
    adapter.clearItems();
    formats = note.getFormats();
    adapter.addItems(formats);

    if (!getEditModeValue()) {
      Set<Tag> tags = note.getTags(context);
      String tagLabel = Note.getTagString(tags);
      if (tagLabel.isEmpty()) {
        return;
      }

      Format format = new Format(FormatType.TAG, tagLabel);
      adapter.addItem(format);
    }
  }

  private void setRecyclerView() {
    adapter = new FormatAdapter(this);
    formatsView = new RecyclerViewBuilder(this)
        .setAdapter(adapter)
        .setView(this, R.id.advanced_note_recycler)
        .build();
  }

  public void setFormat(Format format) {
    // do nothing
  }

  public void moveFormat(int from, int to) {
    // do nothing
  }

  public void deleteFormat(Format format) {
    // do nothing
  }

  public void createOrChangeToNextFormat(Format format) {
    // do nothing
  }

  public void setFormatChecked(Format format, boolean checked) {
    int position = getFormatIndex(format);
    if (position == -1) {
      return;
    }
    format.formatType = checked ? FormatType.CHECKLIST_CHECKED : FormatType.CHECKLIST_UNCHECKED;
    formats.set(position, format);
    adapter.updateItem(format, position);
    updateNote();
  }

  private void setToolbars() {
    toolbar = findViewById(R.id.toolbar);
    colorSelectorLayout = findViewById(R.id.flexbox_layout);
    setButtonToolbar();

    actionDelete = findViewById(R.id.delete_button);
    actionDelete.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        if (note.getNoteState() == NoteState.TRASH) {
          note.delete(context);
        } else {
          note.mark(context, NoteState.TRASH);
        }
        finish();
      }
    });

    actionCopy = findViewById(R.id.copy_button);
    actionCopy.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        note.copy(context);
      }
    });

    backButton = findViewById(R.id.back_button);
    backButton.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        onBackPressed();
      }
    });

    actionOptions = findViewById(R.id.note_options_button);
    actionOptions.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        NoteAdvancedActivityBottomSheet.Companion.openSheet(
            ViewAdvancedNoteActivity.this,
            note,
            getEditModeValue());
      }
    });

    actionShare = findViewById(R.id.share_button);
    actionShare.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        note.share(context);
      }
    });

    actionEdit = findViewById(R.id.edit_button);
    actionEdit.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        openEditor();
      }
    });

    actionDone = findViewById(R.id.done_button);
    actionDone.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        onBackPressed();
      }
    });

    colorButton = findViewById(R.id.color_button);
    setTopToolbar();
    notifyToolbarColor();
  }

  public void openEditor() {
    note.startEditActivity(context);
  }

  protected void notifyToolbarColor() {
    ThemeManager theme =  ThemeManager.Companion.get(this);
    int toolbarIconColor = theme.get(context, ThemeColorType.TOOLBAR_ICON);
    backButton.setColorFilter(toolbarIconColor);
    actionCopy.setColorFilter(toolbarIconColor);
    actionDelete.setColorFilter(toolbarIconColor);
    actionShare.setColorFilter(toolbarIconColor);
    actionEdit.setColorFilter(toolbarIconColor);
    actionDone.setColorFilter(toolbarIconColor);
    actionOptions.setColorFilter(toolbarIconColor);

    int backgroundColor = theme.get(context, ThemeColorType.BACKGROUND);
    rootView.setBackgroundColor(backgroundColor);
    formatsView.setBackgroundColor(backgroundColor);

    resetBundle();
    adapter.notifyDataSetChanged();
    setSystemTheme();
  }

  protected void setButtonToolbar() {
    // do nothing
    toolbar.setVisibility(GONE);
  }

  protected void setTopToolbar() {
    View colorButtonClicker = findViewById(R.id.color_button_clicker);
    colorButtonClicker.setVisibility(GONE);
  }

  protected void setNoteColor(int color) {
    colorButton.setBackground(new CircleDrawable(note.color));
  }

  protected void maybeSaveNoteWithSync() {
    maybeSaveNote(true);
  }

  protected void maybeSaveNote(boolean sync) {
    if (note.getFormats().isEmpty() && note.isUnsaved()) {
      return;
    }
    if (sync) note.save(context);
    else note.saveWithoutSync(context);
  }

  private void updateNote() {
    note.description = Format.getNote(formats);
    maybeSaveNoteWithSync();
  }

  public void moveItemToTrashOrDelete(Note note) {
    if (note.getNoteState() == NoteState.TRASH) {
      note.delete(this);
    } else {
      markItem(note, NoteState.TRASH);
    }
    finish();
  }


  public void markItem(Note note, NoteState state) {
    note.mark(this, state);
  }

  public void notifyNoteChange() {

  }

  public void notifyTagsChanged() {
    setNote();
  }

  protected int getFormatIndex(Format format) {
    int position = 0;
    for (Format fmt : formats) {
      if (fmt.uid == format.uid) {
        return position;
      }
      position++;
    }
    return -1;
  }

  @Override
  public void notifyThemeChange() {
    notifyToolbarColor();
  }

  @Override
  public void onSaveInstanceState(Bundle savedInstanceState) {
    super.onSaveInstanceState(savedInstanceState);
    if (savedInstanceState == null) {
      return;
    }
    savedInstanceState.putInt(NOTE_ID, note == null || note.uid == null ? 0 : note.uid);
  }
}

File: app/src/main/java/com/bijoysingh/quicknote/activities/ViewAdvancedNoteActivity.kt
package com.bijoysingh.quicknote.activities

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.WindowManager
import android.widget.ImageView
import com.bijoysingh.quicknote.MaterialNotes.Companion.appTheme
import com.bijoysingh.quicknote.MaterialNotes.Companion.userPreferences
import com.bijoysingh.quicknote.R
import com.bijoysingh.quicknote.activities.sheets.NoteOptionsBottomSheet
import com.bijoysingh.quicknote.activities.sheets.NoteSettingsOptionsBottomSheet
import com.bijoysingh.quicknote.activities.sheets.SettingsOptionsBottomSheet.Companion.KEY_MARKDOWN_ENABLED
import com.bijoysingh.quicknote.activities.sheets.TextSizeBottomSheet
import com.bijoysingh.quicknote.database.Note
import com.bijoysingh.quicknote.database.utils.*
import com.bijoysingh.quicknote.formats.Format
import com.bijoysingh.quicknote.formats.FormatType
import com.bijoysingh.quicknote.recyclerview.FormatAdapter
import com.bijoysingh.quicknote.recyclerview.FormatTextViewHolder
import com.bijoysingh.quicknote.utils.*
import com.github.bijoysingh.starter.recyclerview.RecyclerViewBuilder
import java.util.*
import kotlin.math.roundToInt


const val INTENT_KEY_NOTE_ID = "NOTE_ID"
const val INTENT_KEY_DISTRACTION_FREE = "DISTRACTION_FREE"

open class ViewAdvancedNoteActivity : ThemedActivity(), INoteOptionSheetActivity {

  var focusedFormat: Format? = null
  protected var note: Note? = null

  protected lateinit var context: Context
  protected lateinit var adapter: FormatAdapter
  protected lateinit var formats: MutableList<Format>
  protected lateinit var formatsView: RecyclerView
  protected var isDistractionFree: Boolean = false

  val topToolbar: View by bind(R.id.top_toolbar_layout)
  val toolbar: View by bind(R.id.toolbar)
  val formatToolbar: View by bind(R.id.format_toolbar)
  val markdownToolbar: View by bind(R.id.markdown_toolbar)

  val rootView: View by bind(R.id.root_layout)
  val backButton: ImageView by bind(R.id.back_button)
  val actionCopy: ImageView by bind(R.id.copy_button)
  val actionDelete: ImageView by bind(R.id.delete_button)
  val actionShare: ImageView by bind(R.id.share_button)
  val actionDone: ImageView by bind(R.id.done_button)
  val colorButton: ImageView by bind(R.id.color_button)

  val primaryFab: FloatingActionButton by bind(R.id.primary_fab_action)
  val secondaryFab: FloatingActionButton by bind(R.id.secondary_fab_action)

  protected open val editModeValue: Boolean
    get() = false

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_advanced_note)
    context = this

    var noteId = intent.getIntExtra(INTENT_KEY_NOTE_ID, 0)
    if (noteId == 0 && savedInstanceState != null) {
      noteId = savedInstanceState.getInt(INTENT_KEY_NOTE_ID, 0)
    }
    if (noteId != 0) {
      note = Note.db().getByID(noteId)
    }
    if (note === null) {
      note = genEmptyNote(NoteSettingsOptionsBottomSheet.genDefaultColor())
    }
    isDistractionFree = intent.getBooleanExtra(INTENT_KEY_DISTRACTION_FREE, false)

    setRecyclerView()
    setToolbars()
    setEditMode()
    notifyThemeChange()
  }

  override fun onResume() {
    super.onResume()
    onResumeAction()
  }

  protected open fun onResumeAction() {
    note = Note.db().getByID(intent.getIntExtra(INTENT_KEY_NOTE_ID, 0))
    if (note == null) {
      finish()
      return
    }
    setNote()
  }

  protected open fun setEditMode() {
    setEditMode(editModeValue)
    formatsView.setBackgroundColor(appTheme().get(ThemeColorType.BACKGROUND))
  }

  protected fun setEditMode(mode: Boolean) {
    resetBundle()
    setNote()

    actionDone.visibility = if (mode) VISIBLE else GONE
    toolbar.visibility = if (mode) VISIBLE else GONE
    primaryFab.visibility = if (mode || isDistractionFree) GONE else VISIBLE
    secondaryFab.visibility = if (mode|| isDistractionFree) GONE else VISIBLE
    markdownToolbar.visibility = GONE
  }

  private fun startDistractionFreeMode() {
    primaryFab.hide()
    secondaryFab.hide()
    topToolbar.visibility = GONE

    var uiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
        or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
        or View.SYSTEM_UI_FLAG_FULLSCREEN)
    if (Build.VERSION.SDK_INT >= 19) {
      uiVisibility = (uiVisibility or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY)
    }
    window.decorView.systemUiVisibility = uiVisibility
  }

  private fun resetBundle() {
    val bundle = Bundle()
    bundle.putBoolean(FormatTextViewHolder.KEY_EDITABLE, editModeValue)
    bundle.putBoolean(KEY_MARKDOWN_ENABLED, userPreferences().get(KEY_MARKDOWN_ENABLED, true))
    bundle.putBoolean(KEY_NIGHT_THEME, appTheme().isNightTheme())
    bundle.putInt(TextSizeBottomSheet.KEY_TEXT_SIZE, TextSizeBottomSheet.getDefaultTextSize())
    bundle.putString(INTENT_KEY_NOTE_ID, note!!.uuid)
    adapter.setExtra(bundle)
  }

  protected open fun setNote() {
    setNoteColor(note!!.color)
    adapter.clearItems()

    if (isDistractionFree) {
      adapter.addItem(Format(FormatType.SEPARATOR))
    }

    formats = note!!.getFormats().toMutableList()
    adapter.addItems(formats)

    if (!editModeValue) {
      maybeAddTags()
      maybeAddEmptySpace()
    }
  }

  private fun maybeAddTags() {
    val tagLabel = note!!.getTagString()
    if (tagLabel.isEmpty()) {
      return
    }

    val format = Format(FormatType.TAG, tagLabel)
    adapter.addItem(format)
  }

  private fun maybeAddEmptySpace() {
    adapter.addItem(Format(FormatType.EMPTY))
  }

  private fun setRecyclerView() {
    adapter = FormatAdapter(this)
    formatsView = RecyclerViewBuilder(this)
        .setAdapter(adapter)
        .setView(this, R.id.advanced_note_recycler)
        .build()
    if (!editModeValue) {
      formatsView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
        private fun onShow() {
          if (isDistractionFree) {
            return
          }
          primaryFab.show()
          secondaryFab.show()
        }

        private fun onHide() {
          primaryFab.hide()
          secondaryFab.hide()
        }

        override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
          super.onScrollStateChanged(recyclerView, newState)
          when (newState) {
            RecyclerView.SCROLL_STATE_DRAGGING -> {
              onHide()
            }
            RecyclerView.SCROLL_STATE_IDLE -> {
              onShow()
            }
          }
        }
      })
    }
  }

  open fun setFormat(format: Format) {
    // do nothing
  }

  open fun moveFormat(fromPosition: Int, toPosition: Int) {
    // do nothing
  }

  open fun deleteFormat(format: Format) {
    // do nothing
  }

  open fun createOrChangeToNextFormat(format: Format) {
    // do nothing
  }

  open fun setFormatChecked(format: Format, checked: Boolean) {
    val position = getFormatIndex(format)
    if (position == -1) {
      return
    }
    format.formatType = if (checked) FormatType.CHECKLIST_CHECKED else FormatType.CHECKLIST_UNCHECKED
    formats[position] = format
    adapter.updateItem(format, position)
    updateNoteForChecked()
  }

  private fun setToolbars() {
    setBottomToolbar()
    actionDelete.setOnClickListener {
      moveItemToTrashOrDelete(note!!)
    }
    actionCopy.setOnClickListener { note!!.copy(context) }
    backButton.setOnClickListener { onBackPressed() }
    actionShare.setOnClickListener { note!!.share(context) }
    actionDone.setOnClickListener { onBackPressed() }
    primaryFab.setOnClickListener { openEditor() }
    secondaryFab.setOnClickListener { openMoreOptions() }
    setTopToolbar()
    notifyToolbarColor()

    if (isDistractionFree) {
      startDistractionFreeMode()
    }
  }

  fun openMoreOptions() {
    NoteOptionsBottomSheet.openSheet(this@ViewAdvancedNoteActivity, note!!)
  }

  fun openEditor() {
    note!!.openEdit(context)
  }

  protected open fun notifyToolbarColor() {
    val theme = appTheme()
    val toolbarIconColor = theme.get(ThemeColorType.TOOLBAR_ICON)
    backButton.setColorFilter(toolbarIconColor)
    actionCopy.setColorFilter(toolbarIconColor)
    actionDelete.setColorFilter(toolbarIconColor)
    actionShare.setColorFilter(toolbarIconColor)
    actionDone.setColorFilter(toolbarIconColor)

    val backgroundColor = theme.get(ThemeColorType.BACKGROUND)
    rootView.setBackgroundColor(backgroundColor)
    formatsView.setBackgroundColor(backgroundColor)

    resetBundle()
    adapter.notifyDataSetChanged()
    setSystemTheme()
  }

  protected open fun setBottomToolbar() {
    toolbar.visibility = GONE
    markdownToolbar.visibility = GONE
  }

  protected open fun setTopToolbar() {
    val colorButtonClicker = findViewById<View>(R.id.color_button_clicker)
    colorButtonClicker.visibility = GONE
  }

  protected open fun setNoteColor(color: Int) {
    colorButton.background = CircleDrawable(note!!.color)
  }

  protected fun maybeSaveNote(sync: Boolean) {
    if (note!!.getFormats().isEmpty() && note!!.isUnsaved()) {
      return
    }
    note!!.updateTimestamp = Calendar.getInstance().timeInMillis
    when (sync) {
      true -> note!!.save(context)
      false -> note!!.saveWithoutSync(context)
    }
  }

  private fun updateNoteForChecked() {
    note!!.description = Format.getNote(formats.sorted())
    setNote()
    maybeSaveNote(true)
  }

  fun notifyNoteChange() {

  }

  protected fun getFormatIndex(format: Format): Int = getFormatIndex(format.uid)

  protected fun getFormatIndex(formatUid: Int): Int {
    var position = 0
    for (fmt in formats) {
      if (fmt.uid == formatUid) {
        return position
      }
      position++
    }
    return -1
  }

  override fun notifyThemeChange() {
    notifyToolbarColor()
  }

  public override fun onSaveInstanceState(savedInstanceState: Bundle?) {
    super.onSaveInstanceState(savedInstanceState)
    if (savedInstanceState == null) {
      return
    }
    savedInstanceState.putInt(INTENT_KEY_NOTE_ID, if (note == null || note!!.uid == null) 0 else note!!.uid)
  }

  companion object {
    const val HANDLER_UPDATE_TIME = 1000

    fun getIntent(context: Context, note: Note): Intent {
      val intent = Intent(context, ViewAdvancedNoteActivity::class.java)
      intent.putExtra(INTENT_KEY_NOTE_ID, note.uid)
      return intent
    }
  }

  /**
   * Start : INoteOptionSheetActivity Functions
   */

  override fun updateNote(note: Note) {
    note.save(this)
    notifyNoteChange()
  }

  override fun markItem(note: Note, state: NoteState) {
    note.mark(this, state)
  }

  override fun moveItemToTrashOrDelete(note: Note) {
    note.deleteOrMoveToTrash(context)
    finish()
  }

  override fun notifyTagsChanged(note: Note) {
    setNote()
  }

  override fun getSelectMode(note: Note): String {
    return NoteState.DEFAULT.name
  }

  override fun notifyResetOrDismiss() {
    finish()
  }

  override fun lockedContentIsHidden() = false

  /**
   * End : INoteOptionSheetActivity
   */

}
